contracts/exchangeIssuance/IssueRedeem.sol

```diff
@@ -0,0 +1,192 @@
+        address[] memory path = new address[](2);
+        path[0] = WETH;
+        path[1] = _token;
+        _router.swapTokensForExactTokens(_amount, MAX_INT, path, address(this), block.timestamp);
+    }
+
+    function purchaseTokenBestPrice(address _token, uint256 _amount) private {
+        uint256 uniPrice = tokenAvailable(uniFactory, _token) ? getBuyPrice(true, _token, _amount) : MAX_INT;
+        uint256 sushiPrice = tokenAvailable(sushiFactory, _token) ? getBuyPrice(false, _token, _amount) : MAX_INT;
+        if (uniPrice <= sushiPrice) {
+            purchaseToken(uniRouter, _token, _amount);
+        } else {
+            purchaseToken(sushiRouter, _token, _amount);
+        }
+    }
+
+    function sellToken(IUniswapV2Router02 _router, address _token) private {
+        uint256 tokenBalance = IERC20(_token).balanceOf(address(this));
+        address[] memory path = new address[](2);
+        path[0] = _token;
+        path[1] = WETH;
+        _router.swapExactTokensForETH(tokenBalance, 0, path, msg.sender, block.timestamp);
+    }
+
+    function sellTokenBestPrice(address _token) private {
+        uint256 tokenBalance = IERC20(_token).balanceOf(address(this));
+        uint256 uniPrice = tokenAvailable(uniFactory, _token) ? getSellPrice(true, _token, tokenBalance) : 0;
+        uint256 sushiPrice = tokenAvailable(sushiFactory, _token) ? getSellPrice(false, _token, tokenBalance) : 0;
+        if (uniPrice >= sushiPrice) {
+            sellToken(uniRouter, _token);
+        } else {
+            sellToken(sushiRouter, _token);
+        }
+    }
+
+    function getBuyPrice(bool isUni, address _token, uint256 _amount) private view returns (uint256) {
+        if (isUni) {
+            (uint256 tokenReserveA, uint256 tokenReserveB) = UniswapV2Library.getReserves(uniFactory, WETH, _token);
+            return uniRouter.getAmountIn(_amount, tokenReserveA, tokenReserveB);
+        } else {
+            (uint256 tokenReserveA, uint256 tokenReserveB) = SushiswapV2Library.getReserves(sushiFactory, WETH, _token);
+            return sushiRouter.getAmountIn(_amount, tokenReserveA, tokenReserveB);
+        }
+    }
+
+    function getSellPrice(bool isUni, address _token, uint256 _amount) private view returns (uint256) {
+        if (isUni) {
+            (uint256 tokenReserveA, uint256 tokenReserveB) = UniswapV2Library.getReserves(uniFactory, WETH, _token);
+            return uniRouter.getAmountOut(_amount, tokenReserveB, tokenReserveA);
+        } else {
+            (uint256 tokenReserveA, uint256 tokenReserveB) = SushiswapV2Library.getReserves(sushiFactory, WETH, _token);
+            return sushiRouter.getAmountOut(_amount, tokenReserveB, tokenReserveA);
+        }
+    }
+
+    function tokenAvailable(address _factory, address _token) private view returns (bool) {
+        return IUniswapV2Factory(_factory).getPair(WETH, _token) != address(0);
+    }
+
+    function preciseMulCeil(uint256 a, uint256 b) internal pure returns (uint256) {
```

Comment:
Instead of rewriting the function, import the correct preciseMulCeil from

https://github.com/SetProtocol/index-coop-contracts/blob/master/contracts/lib/PreciseUnitMath.sol